# Práctica Clase 1 

1) Crear una función capaz de convertir números enteros de base 10 a base 2. Debe recibir como parámetro el número a convertir<br>
Consideraciones:<br> 
a. Tratar de resolverlo sin usar la función format(nro,"b")<br>
b. El pdf "conversion-de-decimal-a-binario.pdf" puede servir de ayuda.


In [52]:
def NumeroBinario(numero):
    '''
    Esta función recibe como parámetro un número entero mayor ó igual a cero y lo devuelve en su 
    representación binaria. Debe recibir y devolver un valor de tipo entero.
    En caso de que el parámetro no sea de tipo entero y mayor a -1 retorna nulo.
    '''
    if type(numero) != int or numero < 0:
        return None
    elif (numero == 0):
        return 0
    lisbin = []
    resto = 0

    while (numero > 0):
        resto = numero % 2
        numero = numero // 2
        lisbin.append(resto)

    binario = ''
    for e in lisbin:
        binario = str(e) + binario
    
    return int(binario)


In [70]:
def NumeroBinario(numero):
    '''
    otra opcion
    '''
    if type(numero) != int or numero < 0:
        return None
    elif (numero == 0):
        return 0
    binario = ''
    resto = 0

    while (numero > 0):
        resto = numero % 2
        numero //= 2
        binario += str(resto)
  
     
    return int(binario[::-1])


In [73]:
#Ejemplo:
NumeroBinario(13)

1101

2) Convertir de decimal a binario las fracciones 1/2, 1/3, 1/4, 1/5, 1/6, 1/7, 1/8, 1/9. Luego analizar los resultados y observar qué particularidad se encuentra en los mismos. Se puede usar Python o una calculadora, lo importante es ver si hay algo que podemos notar...

In [55]:
def Binafrac(numdec):
    if type (numdec) != float or  numdec < 0 or numdec > 1:
        return None
    elif (numdec == 0):
        return 0

    limite = 24
    listabin = []
    i=0
    while (numdec > 0) and (i < limite):
        listabin.append(int(numdec * 2))
        numdec = numdec * 2 - int(numdec * 2)
        i+=1

    numbin = '0.'
    for e in listabin:
        numbin = numbin + str(e)
    return numbin



In [74]:
print(Binafrac(13))


None


In [75]:
bin(13)

'0b1101'

In [42]:
for i in range(2,10):
    print('Fraccion 1 /', i, ':', str(1/i), ' y en binario :', Binafrac(1/i))

Fraccion 1 / 2 : 0.5  y en binario : 0.1
Fraccion 1 / 3 : 0.3333333333333333  y en binario : 0.01010101010101010101
Fraccion 1 / 4 : 0.25  y en binario : 0.01
Fraccion 1 / 5 : 0.2  y en binario : 0.00110011001100110011
Fraccion 1 / 6 : 0.16666666666666666  y en binario : 0.00101010101010101010
Fraccion 1 / 7 : 0.14285714285714285  y en binario : 0.00100100100100100100
Fraccion 1 / 8 : 0.125  y en binario : 0.001
Fraccion 1 / 9 : 0.1111111111111111  y en binario : 0.00011100011100011100


Salida esperada:
* Fraccion 1 / 2 :  0.5  y En binario:  0.1
* Fraccion 1 / 3 :  0.3333333333333333  y En binario:  0.010101010101010101010101
* Fraccion 1 / 4 :  0.25  y En binario:  0.01
* Fraccion 1 / 5 :  0.2  y En binario:  0.001100110011001100110011
* Fraccion 1 / 6 :  0.16666666666666666  y En binario:  0.001010101010101010101010
* Fraccion 1 / 7 :  0.14285714285714285  y En binario:  0.001001001001001001001001
* Fraccion 1 / 8 :  0.125  y En binario:  0.001
* Fraccion 1 / 9 :  0.1111111111111111  y En binario:  0.000111000111000111000111